src/foam/nanos/crunch/README.md

```diff
@@ -0,0 +1,83 @@
+
+##### Capability - (Prerequisite) Capability
+- Stored in the `prerequisiteCapabilityJunctionDAO`
+- If a Capability has a prerequisite Capability, the user must be GRANTED the prerequisite Capability before being eligible to apply for the Capability
+
+##### Capability - (Deprecated) Capability
+- stored in the `deprecatedCapabilityJunctionDAO`
+- If a Capability has been deprecated by another Capability, it is essentially no longer enabled, and ignored by the system
+
+##### User - Capability 
+- stored in the `userCapabilityJunctionDAO`
+- refined by `UserCapabilityJunctionRefine`
+Notable properties on the UCJ model :
+- `status` - an Enum of CapabilityJunctionStatus, `PENDING` unless `GRANTED` or `EXPIRED`
+- `expiry` - A DateTime calculated at the time the UCJ becomes GRANTED
+- `data` - Information relating to the user and the capability. 
+    - In the liquid use case, this field was used to add information about the extend of the Capability granted to the user. 
+    - In the onboarding use case, this could be information that need be collected from the user in order to grant a Capability
+The UserCapabilityJunctionDAO decorator :
+- This decorator was created with the onboarding use case in mind and can be excluded depending on need
+- Behavior on put :
+    - ownership of the UCJ is checked
+    - UCJs with prerequisite Capabilities are checked
+    - the `data` of the UCJ must be a validatable, and its `validate` method is checked 
+    - the `data` is saved to the `daoKey` of the Capability, if `daoKey` is provided
+    - if the prerequistes are fulfilled and the data is validated, the `status` of the UCJ is set to `GRANTED`, and the `expiry` is calculated. Otherwise, the `status` is set to `PENDING`
+
+#### CapabilityJunctionStatus
+Includes : 
+- PENDING
+- GRANTED
+- EXPIRED
+
+#### CapabilityAuthService
+Overrides the `checkUser` method: Given a permission `p`, and User `u`,
+1. Check whether `p` is the name of an enabled and non-deprecated Capability
+2. Try to find a GRANTED and non-EXPIRED UserCapabilityJunction such that `User == u && Capability == p`. If found, return true
+3. Find the non-EXPIRED UserCapabilityJunctions where `User == u`, for each UCJ, check if the Capability is non-DEPRECATED and implies `p`
+    - `implies` is a method on the Capability model. A Capability `c` implies a String `p` if :
+        - `c` has a prerequisite Capability which implies `p`
+        - `p` is in the set of `permissionsGranted` by `c`, or implied by the set of `permissionsGranted` by `c`*.
+        - the name of `c` implies `p`* 
+        - *e.g., `obj.*` implies `obj.read`
+
+A notable difference between this decorator and other AuthService decorators is that if at any point, the result of the check is found to be true, `true` is returned instead of moving on to the next decorators in the AuthService chain. And if at any point, the result of the check is found to be false, the check moves on to return the result of its delegate
```

Comment:
Is this really different?

Reply:
I would say that while this AuthService can grant permissions to a user, it won't remove permissions that they receive from other AuthServices in the chain.

Reply:
I think other decorators that override check does the opposite of this where it goes on to check the result of its delegates when the result of the current check is found to be true, and returns false when the result of current is found to be false. In this case, we are returning true on true and returning the delegate on false since the delegate will not return true for crunch-specific permission strings.